src/core/thread/csl_tx_scheduler.cpp

```diff
@@ -158,9 +158,10 @@ uint32_t CslTxScheduler::GetNextCslTransmissionDelay(const Child &aChild,
 {
     uint64_t radioNow   = otPlatRadioGetNow(&GetInstance());
     uint32_t periodInUs = aChild.GetCslPeriod() * kUsPerTenSymbols;
-    uint64_t firstTxWindow =
-        aChild.GetLastRxTimestamp() - kRadioHeaderShrDuration + aChild.GetCslPhase() * kUsPerTenSymbols;
-    uint64_t nextTxWindow = radioNow - (radioNow % periodInUs) + (firstTxWindow % periodInUs);
+
+    /* see CslTxScheduler::ChildInfo::mCslPhase */
+    uint64_t firstTxWindow = aChild.GetLastRxTimestamp() + aChild.GetCslPhase() * kUsPerTenSymbols;
```

Comment:
@EskoDijk Thank you for looking into the change at such a level of detail! This ensures that we don't overlook anything and that everyone understands what this implies. Let me try to answer your questions as precisely as I can.

> Is it intentional that kRadioHeaderShrDuration is removed here -- so that the point of reference for nextTxWindow becomes "end of last symbol of SFD" ?

Moving to a common standard based definition for TX and RX timestamps was the central motivation for this PR as those definitions were inconsistent in OT AFAICS (see detailed rationale of my choice of timestamp definition below).

> Because this changes the calculation, it also means that code below in sub_mac.cpp becomes invalid:

I don't think so. The change was introduced asymmetrically as the TX side was already based on the SFD definition (at least in Zepyhr's platform adaptation layer). I may have done it wrong or my memory may fail me - but that was the intention. @edmont You're probably in a much better place to explain the details on the OT side or see the errors in my implementation or argument.

> Maybe the change on the txWindow calculations can be done in a separate PR? As it modifies the calculation and breaks any platforms that use a non-time-scheduled Tx platform.

All the documentation is just preparative to introduce the concepts needed to define timestamps with the precision required by the IEEE 802.15.4 standard itself. The documentation describes and explains the implementation change and can therefore not be separated from it IMHO. I just did two commits to ensure that the implementation change is not "lost in the noise". Of course I may be wrong or the documentation may not reflect the implementation - so I hope such errors will be unveiled in this review.

This PR proposes a common definition for TX and RX timestamps for the following reasons:
* Such an inconsistency is difficult to reason about and easy to overlook. My intention is to reduce cognitive load and improve maintainability (better long term TCO at the price of short term refactoring effort).
* There is also a technical argument that might not (yet) be relevant to OpenThread but this can easily change in the future (as the Thread standard may not ignore highly relevant recent PHYs like UWB or SUN FSK forever). The IEEE 802.15.4 standard defines a "start of frame delimiter" that it guarantees to be consistently defined across all current and future PHYs such that MAC protocols can be based on that definition regardless of PHY details. SFD is defined as the earliest point in time at which the existence of a frame on the medium can be predicted with sufficiently high probablility and precision (including ranging requirements) for a given RF demodulation/correlation/filter technology. The SHR duration depends very much on the PHY and can therefore not be consistently used on MAC level. AFAIK this is the reason why most frame timings in the standard are relative to the "start of the first symbol after the SFD" or "end of SFD". If even more precision is required (e.g. in ranging) then the timestamp reference plane is defined to be "the local antenna" and called "RMARKER" (but it's the same basic concept). The fact that CSL also introduces timings relative to the start of the MAC header is probably to ease the implementation of the corresponding IE parsers and as you can see from my formula is not a problem as long as the PHY specific offsets cancel. Such a definition is not adequate for general timestamps independent of the IEEE 802.15.4 sub-protocol, though, and should therefore not be used outside CSL standard for general RX/TX timestamps.

I agree that it is unfortunate to fix such an inconsistency after it has been implemented by all platforms. But OTOH the longer it lives the harder it will be to fix. So IMO it's still cheaper to fix it now than in two years.

IMO the change is less than what you assume, though. It's just that the length of the SHR (which is just a simple well-defined us-precision constant for the given PHY) may have to be added in platform implementations that choose to use different timestamp definitions internally. In Zephyr we simply migrated our whole stack down to the driver level to a standard definition - but that's our choice of course. I think it's obvious that this is the right thing to do for Zephyr (which already supports all kinds of PHYs and IEEE 802.15.4 standard protocols and expects to do some work in ranging soon).

That said, I'm just offering my help here. Zephyr's decision to use standard timestamps can easily be isolated from the OT API by transforming timestamps internally to Zephyr's OT platform adaptation layer. So of course the OT community may legitimately decide to use different timestamp definition for whatever theoretical or practical reasons (which would be a valid result of the review!).

Reply:
@EskoDijk Fixed the SHR offset for platforms that don't support timed TX as you suggested.

Reply:
@fgrandel Ok thanks for pointing out the motivation. Most other (active) participants I think agreed with the idea to unify timestamp references. But I'm not sure that a one-type-of-time-reference approach is possible. Below some thoughts about that.  Maybe we could agree that any variable called "timestamp" uses the end-of-last-symbol-of-SFD time reference, while any variable using "window" (win) in its name uses the "when the radio is turned on or off" time reference?  If agreed in this way then I can check details of the code to see if it's all correct. And I can verify against my simulation platform to check that the CSL timings are correctly calculated.

Thoughts:

The header `include/openthread/platform/radio.h` introduces also the concept of "window" in `otPlatRadioReceiveAt` and this window concept comes back in other parts of code like `GetNextCslTransmissionDelay` and `SubMac::HandleCslTimer(void)`  (abbreviated 'win' there).  So the concept is used for the receiver (time that radio is on, see e.g. fig 6-63 in 802.15.4-2015) and also for the transmitter - although for that case it refers more to a single timestamp, that targets a time that lies within the receiver's window.

This relates directly to the 802.15.4 concept of the "channel sample". The start of this channel sample (period) is called channel sample time, and that is a timestamp that refers to the time that the radio is turned on (i.e. referencing the start of first symbol of SHR, from the receiver's perspective.) 
So in my mind the "channel sample" as defined by 802.15.4 is equivalent to the "window" that OpenThread introduces for the receiver. 

In my mind, to remain well aligned to the standards, we need to keep the same concept and not change it to "Rx window start is end of last symbol of SFD".  That would be more confusing I think. If that reasoning is correct we have to live with multiple methods of time reference, because the standard itself is also doing this.

One question here: I'm not sure if the current PR aims to change the "receive window start time" (`aStart`) parameter reference time, in `otPlatRadioReceiveAt`, from its current time interpretation of "start of window" to "end of last symbol of SFD" ?  If yes that would imply some more code changes are needed in the calls to `otPlatRadioReceiveAt`.

At the moment, on the CSL receiver, in `SubMac::HandleCslTimer(void)` the variable `winStart` is not a timestamp that is pointing to the end of last symbol of SFD. Rather it points to the "start of the window - when the CSL receiver switches on the radio".
The reason for this is simple: the code for platforms-without-receive-time-support needs to switch on the radio in Rx at this time so it's not so handy or intuitive I think to use another timestamp reference. The timer needs to trigger at the time that the radio must be switched on; and not later than that. And "window start" refers to the 802.15.4 concept of "channel sample" i.e. start of the channel sample.

Reply:
> In my mind, to remain well aligned to the standards, we need to keep the same concept and not change it to "Rx window start is end of last symbol of SFD". 

@EskoDijk Agreed! (This and everything else you write in that comment)

I never intended to say anything about RX windows or channel samples in this PR other than documenting existing realities or pointing out the simple change in offset required when deriving the existing RX windows from a slightly changed TX timestamp. I also agree that SFD is relevant for timestamping only.

The only thing that I intend to change in this PR is the TX timestamp because it deviates from the standard AND from the RX timestamp definition. (Timestamps as opposed to RX windows exactly as you define them.) If I did say anything that results in a change to the resulting RX window definition, it is unintended and should be pointed out and fixed.

Reply:
Ok thanks, good to hear - so we can keep the "window" term for the receiver's channel sample then I think; and also move the timestamps to the agreed reference point. And the CSL timer can trigger in advance, as needed for the specific platform. With this in mind I can do another review and then test with the simulation platform! (hopefully by tomorrow)